cpp/include/cudf/groupby.hpp

```diff
@@ -43,6 +52,58 @@ struct aggregation_request {
       aggregations;  ///< Desired aggregations
 };
 
+/**
+ * @brief Struct defining the bounds of a window-based analytical query.
+ */
+struct window_bounds {
+
+  // Number of rows preceding the current row, to consider in the window
+  cudf::size_type preceding;    
+
+  // Number of rows following the current row, to consider in the window
+  cudf::size_type following; 
+
+  // The minimum number of observations to consider, for a window to be
+  // valid, including: 
+  // 1. The current row
+  // 2. Upto `preceding` rows appearing prior to the current row
+  //    (subject to belonging within a group)
+  // 3. Upto `following` rows appearing after the current row
+  //    (subject to belonging within a group)
+  // 
+  // E.g. Consider the following int column-vector:
+  //   [0,1,2,3,4,5]
+  // If an aggregation operation (E.g. SUM) is specified with
+  // the following window parameters:
+  //   preceding   = 1
+  //   following   = 1
+  //   min_periods = 3
+  //
+  // Then, the resulting column-vector would be as follows:
+  //   [INVALID, 3, 6, 9, 12, INVALID]
+  // 
+  // Specifically, the results at indices `0` and `5` are INVALID, 
+  // since there are only two observations (each) at either end,
+  // and *three* are needed at minimum, for the result to be valid.
+  cudf::size_type min_periods;
+};
+
+/**
+ * @brief Request for groupby window-based aggregation(s) to perform on a column.
+ * 
+ * Similarly to `aggregation_request`, a `window_aggregation_request` is applied
+ * to all values in the column vector, but only across the elements within the 
+ * confines of the associated `window_bounds`.
+ * 
+ * `values.size()` must equal `keys.num_rows()`.
+ */
+struct window_aggregation_request {
+  column_view values; ///< the elements to aggregate
+  std::vector<
+    std::pair<window_bounds, std::unique_ptr<aggregation>>> 
```

Comment:
I'd've preferred to make these aggregations work with `result_cache`. Right now, rolling supports `mean` in a single pass but in the future, adding more aggregations with dependency to rolling should use the cache. Then, we'd need to encode the window information in the cache key. Because a `std` with a certain window would require sum, mean and count with the same window. We can achieve this in two ways:
1. Instantiate the cache to accept `window_bounds`. This can be done by making `result_cache` templated.
2. Change the aggregation objects accepted by this API to new `windowed_<agg>_aggregation` e.g. `windowed_std_aggregation` that inherits from `std_aggregation` and has `window_bounds` as a member. And then, add an `operator==` to it that also checks for `window_bounds`.